server.go

```diff
@@ -3195,7 +3195,13 @@ func (s *server) NotifyWhenOnline(peerKey [33]byte,
 		select {
 		case <-peer.ActiveSignal():
 		case <-peer.QuitSignal():
-			// The peer quit so we'll just return.
+			// The peer quit so we'll add the channel to the slice
+			// and return.
+			s.mu.Lock()
+			s.peerConnectedListeners[pubStr] = append(
```

Comment:
Nice catch! I think this might also be affecting reliable direct sends of channel update messages at times.

Reply:
I think it'd be ideal for us to also add a unit test here. Most of the state here can be pretty easily subbed out (https://github.com/lightningnetwork/lnd/pull/5700 should make this easier in the future). 

So in this case, the test would add a canned peer to `peersByPub` (which doesn't seem to use the main peer interface rn actually), then fire either of the signals to make it enter this new case. A follow up block would then just call `peerInitializer` which'll then send the signals as normal.

Thoughts?

Reply:
I don't think this can be done easily - `peerInitializer` needs a server object and eventually calls `p.Start()` which requires a channeldb and launches goroutines that require other things like a mocked connection. It would be easier if `peersByPub` used an interface instead of `peer.Brontide` which would let us precisely control this. Another issue is that any time `peer.Brontide` is started successfully, the `ActiveSignal` is used, meaning that this test using `peerInitializer` will have both the `ActiveSignal` and the `QuitSignal` active in that select statement, so it could lead to not exercising the quit logic all the time

Reply:
#5700 doesn't change `peersByPub` and the other related maps to an interface, but I think that would be the way to go to test this.

Also independent of this bug fix, some of the callsites of `NotifyWhenOnline` don't retry if a later call to `SendMessage` failed. This is problematic since the peer could have gone down by the time you end up sending the message.